### DISCUSSION TITLE
How to codesign bottles in tap?

### DISCUSSION THREAD

**Discussion by ohueter** [Homebrew/discussions#3614] [Tap maintenance and Homebrew development]
Hi,

I'm maintaining the free [autokbisw](https://github.com/ohueter/autokbisw) app, which is a MacOS service that automatically switches keyboard layouts depending on the physical keyboard used. The app is available from Homebrew via my tap at https://github.com/ohueter/homebrew-tap/blob/master/Formula/autokbisw.rb.

I wanted to supply bottles from this tap, and got seriously confused on how to set this up. 

In the end, I managed to build binaries for Monterey and Big Sur (each amd64 and x86_64) in GitHub Workflows in the `autokbisw` repo and add them to the releases with filenames that Homebrew expects ([latest release here](https://github.com/ohueter/autokbisw/releases/tag/1.4.3)). I also (manually) added the `bottles do` section to the Formula, and Homebrew would pick up the bottles. So far, so good.

However, when using one of these bottles, MacOS would complain because the app couldn't be checked for malware - I guess I need to codesign the binaries? I have no experience with this, is this even possible when compiling on a GitHub-hosted runner? Would using the test-bot workflow help with this? 

I have hardly any experience with native MacOS development, I became the maintainer after the original author retired. I would greatly appreciate if someone could point me in the right direction. Thanks a lot!

**Comment by SMillerDev** [comment]
It's really hard to give advice without seeing the actual errors. But seeing how the test-bot flow is what powers homebrew-core, that seems like the way to go for distributing formulae.

**Reply by ohueter** [reply]
Thanks for your response. I admit my question was very broad, let me try to narrow it down a bit.

The current situation is, when you download one of the built binaries and try to run it, MacOS will display the following warning/error (generic screenshot):
<img src="https://user-images.githubusercontent.com/57348167/188859290-69b75cf9-cff6-40f4-b151-2771b2fd7ba5.png" width="50%">

One possible solution to fix this, as far as I understand, would require me to buy into the Apple developer program and somehow codesign the binaries myself.

However, the official Homebrew bottles don't trigger this error, so they are likely signed somewhere along their way.

My questions are: Is this official Homebrew codesigning mechanism somehow available to private taps? Are there other ways to prevent this warning from appearing?

**Reply by SMillerDev** [reply]
What are you trying to run when you get this warning?

**Reply by ohueter** [reply]
One of the binaries attached to this release: https://github.com/ohueter/autokbisw/releases/tag/1.4.3

In fact, this message appears when you run it by double-clicking in Finder. When executing from the command line, the warning is different:
<img src="https://user-images.githubusercontent.com/57348167/189058637-407ebf52-9d59-4d08-a22b-292477901f38.png" width="50%">

(I'm not attaching own screenshots because my system language isn't English)

**Reply by SMillerDev** [reply]
So is this an app file? Because as far as I know only binaries get codesigned.

**Reply by ohueter** [reply]
No, it's a binary single executable file.

**Comment by carlocab** [comment]
You could try adding a post-install to your formula that does something like
```ruby
def post_install
  system "codesign", "--sign", "-", bin/"autokbisw"
end
```
to see if it helps.

**Reply by ohueter** [reply]
Thanks, that's a great idea, however the command throws an error (I ran it in Terminal):
```
$ codesign --sign autokbisw 
error: The specified item could not be found in the keychain.
```

I guess because of my missing developer certificate ... ?

Don't get me wrong, I don't want to abuse this discussion to get help with code signing itself, I'm just curious if Homebrew has any mechanisms that would help me in this situation without requiring me to buy a developer certificate.

UPDATE:
Sorry, I missed the dash before the filename! Running `codesign --sign - autokbisw` returns no error, but the MacOS warning message remains ...

**Reply by carlocab** [reply]
What does
```
codesign -vvv /path/to/autokbisw
```
say after installing from a bottle with the above `post_install` method?

**Reply by ohueter** [reply]
```
$ codesign -vvv /opt/homebrew/Cellar/autokbisw/1.4.3/bin/autokbisw
autokbisw: valid on disk
autokbisw: satisfies its Designated Requirement
```

**Reply by SMillerDev** [reply]
If you still get the error after that, what is the output of `which -a autokbisw`

**Reply by ohueter** [reply]
```
 $ which -a autokbisw
/opt/homebrew/bin/autokbisw
```
which is a symlink to the path from above
```
 $ ls -la /opt/homebrew/bin
lrwxr-xr-x    1 ole  admin    39B Sep  8 14:07 autokbisw -> ../Cellar/autokbisw/1.4.3/bin/autokbisw
```

What is the expected effect of running `codesign --sign - autokbisw` post install? Should it be equivalent to the user using Ctrl+Click in Finder to override the warning when starting the executable?

**Reply by SMillerDev** [reply]
The effect is that it signs the binary so then you shouldn't have the popup.

**Reply by ohueter** [reply]
As I said, there is no developer certificate installed on the GitHub runners nor my machine, so I don't see how that could work? I was hoping that the test-bot Workflow would contain a generic certificate from the Homebrew maintainers (or create a self-signed certificate during the process) that could be used.

**Reply by carlocab** [reply]
The only way to sign with a "generic certificate" is to do `codesign --sign -`. (See `SIGNING IDENTITIES` in `man codesign`.)

This is also the only type of `codesign`ing ever done by `brew`. See, for example: https://github.com/Homebrew/brew/blob/a73a1a665e9d6cb1d51c714a787ff823efa8b84a/Library/Homebrew/extend/os/mac/keg.rb#L36

If you've applied an ad-hoc signature with `codesign --sign -` and it still doesn't work, then I suspect that your issue isn't that the binary lacks a signature.

**Reply by ohueter** [reply]
That's interesting, I'll look into this. Thank you both for your help, and for pointing me into this direction!

**Reply by Mactastic1-5** [reply]
Are we allowed to add `codesign --force --deep -s - /path/to/app` to casks?

**Comment by ohueter** [comment]
I gave the test-bot workflows another try and got them running. They are convenient, but don't seem to have any effect on code signing the binary. Also cross-compiling for `arm64` does not seem to be possible with them, although the compiler would support it. I guess I'll wait for the repo to collect some more stars and then try to get it into `core`. Until then, the users will need to have XCode installed :-)

**Reply by lukaso** [reply]
If you want to do it on your own, here's some code that does code signing for GIMP on the Mac. Look at 

https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/blob/master/.circleci/config.yml#L185-195
https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/blob/master/package/build.sh#L133-151
https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/blob/master/package/gimp-hardening.entitlements
https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/blob/master/package/notarize.sh

You'll also need an Apple Developer account and then get the right credentials.
